### PR TITLE
Add schema documentation (and slight update to schema)

### PR DESCRIPTION
**Pull request by rasky**
I wrote an initial draft of the complete schema documentation, and did some slight updates to the schema itself.

### CODE REVIEW

File: schema.md [hunk 1]
```diff
@@ -0,0 +1,555 @@
...
+* Presence: optional
+* Example: "NFC Reader (only for scanning tags)"
+
+This key contains a list of hardware dependencies that must be owned
+to use the software.
+
+Each string is free form, max 50 characters; feel free to add version
+numbers and/or short comments.
+
+
```
Comment (bvhme) [inline schema.md:534]: Trim empty lines plz :)

File: schema.md [hunk 2]
```diff
@@ -0,0 +1,555 @@
...
+* Italy: [Italian extensions](it-schema.md)
+
+
+## Top-level section
+
+### Key `version`
+
+* Type: string
+* Presence: mandatory
+* Example: `"0.1"`
```
Comment (bvhme) [inline schema.md:39]: If we want to SEMVER this then this should be

```yaml
version: "0.1.0"
```

(Also the comment in the example still reads `0.0.1`)

File: schema.md [hunk 3]
```diff
@@ -0,0 +1,555 @@
...
+which this fork was created. This key is meant to create a relation
+between forked repositories that could not be found out otherwise (for
+instance, some code hosting platforms record fork relations in their
+metadata).
+
+If you fork a software with the purpose of creating an independent
+version, you should consider mentioning the upstream project in this
+key.
+
+
```
Comment (bvhme) [inline schema.md]: Trim superfluous whitespace and linebreaks.

File: schema.md [hunk 4]
```diff
@@ -0,0 +1,555 @@
...
+* Type: string
+* Presence: mandatory
+* Example: `"AGPL-3.0-or-later"`
+
+This string describes the license under which the software is
+distributed. The string must contain a valid SPDX expression, referring
+to one (or multiple) open-source license. Please refer to the [SPDX
+documentation](https://spdx.org/licenses/) for further information.
+
+### Key `legal/main-copyright-owner`
```
Comment (bvhme) [inline schema.md]: I think we can omit the `main-` from this key. Also I think there is no problem dropping the `-owner` as well (although it makes more sense with the `repo-owner` below). Also consider a `homepage` as addition for this.

File: schema.md [hunk 5]
```diff
@@ -0,0 +1,555 @@
...
+
+This section provides information on the maintenance status of the
+software, useful to evaluate whether the software is actively
+developed or not.
+
+### Key `maintenance/type`
+
+* Type: enumerate
+* Presence: mandatory
+* Values: "commercial", "community", "none"
```
Comment (bvhme) [inline schema.md]: Can I suggest:

Values:
* `internal`: maintenance is performed by the authors
* `contract`: maintenance is performed by a contracted party
* `community`: maintenance is done by the community
* `none`: this project is unmaintained, if you file a pull request you'll likely get no response

File: schema.md [hunk 6]
```diff
@@ -0,0 +1,555 @@
...
+"Commercial" means that there is a commercial contract that binds
+an entity to the maintenance of the software; "community" means that
+the software is currently maintained by one or more people that donate
+their time to the project; "none" means that the software is not
+actively maintained.
+
+### Key `maintenance/until`
+
+* Type: date (YYYY-MM-DD)
+* Presence: mandatory (if the software is commercially maintained)
```
Comment (bvhme) [inline schema.md]: I would say mandatory for any project that is budgetted, ideally we can also see from PA software when the budget is running out.

File: schema.md [hunk 7]
```diff
@@ -0,0 +1,555 @@
...
+* Example: "Linus Torvalds", "Italian Linux Group", "Acme Inc."
+
+This key describes the entity or entities that are currently maintaining
+the software. It can contain for instance a company name, an association
+name, a private person. etc.
+
+In case of a commercial agreement (or a chain of such agreements),
+specify the final entities actually contracted to deliver the
+maintenance. Do not specify the software owner unless it is technically
+involved with the maintenance of the product as well.
```
Comment (bvhme) [inline schema.md:177]: Also, include a `homepage` here I think, as it enables a way to track a maintainer's projects across spellings.

File: example/publiccode.yml [hunk 8]
```diff
@@ -1,28 +1,34 @@
-version: 0.0.1                     # publiccode YAML standard version 0.0.1
+version: "0.1" /                    # publiccode YAML standard version 0.0.1
```
Comment (bvhme) [inline example/publiccode.yml]: The comment still reads `0.0.1` (also semver should be `0.1.0`

File: example/publiccode.yml [hunk 9]
```diff
@@ -1,28 +1,34 @@
...
-    - name: Dario Bianchi
-      email: dario.bianchi@fornitore.it
-      affiliation: Fornitore Privato Spa
+maintenance:                        # Maintainance status as open-source software
+  type: commercial                  # Whether there is a commercial contract of maintainance in place. No -> community
+  until: "2019-01-01"               # If maintainance-type is commercial, then the maintenance contract is due
+  maintainer: Company Spa           # Name of the lead maintainer
+  technical-contacts:               # Name of the people that are currently technically maintaining this software
+  - name: Francesco Rossi
+    email: "francesco.rossi@comune.reggioemilia.it"
```
Comment (bvhme) [inline example/publiccode.yml]: Is there a reason emails are `"` strings and the rest not?

File: schema.md [hunk 10]
```diff
@@ -0,0 +1,555 @@
...
+
+* Type: array of objects
+* Presence: mandatory (if there is a maintenance)
+
+This key describes the physical point of contacts of the technical
+people in charge of the maintenance. The goal is to provide a clear
+contact point for contributors that need to discuss how to best
+implement modifications to the software.
+
+It MUST contain real people names (first name, family name) and direct
```
Comment (bvhme) [inline schema.md]: In some countries the default is `(family name, given name)`, and perhaps that is okay as well. I would rephrase as  `real names of individuals`

File: schema.md [hunk 11]
```diff
@@ -0,0 +1,555 @@
...
+
+This key contains the latest stable version number of the software.
+The version number is a string that is not meant to be interpreted
+and parsed but just displayed; parsers should not assume semantic
+versioning or any other specific version format.
+
+The key can be omitted if the software is currently in initial
+development and has never been released yet.
+
+### Key `description/released`
```
Comment (bvhme) [inline schema.md]: Perhaps `release-date` is clearer?

File: schema.md [hunk 12]
```diff
@@ -0,0 +1,555 @@
...
+Use the predefined values if possible. If the software runs on a
+platform for which a predefined value is not available, a different
+value can be used.
+
+### Key `description/logo`
+
+* Type: string or array of strings (path to file)
+* Presence: optional
+* Formats: SVG, PNG
+* Example: "img/logo.svg"
```
Comment (bvhme) [inline schema.md:300]: I would make this an absolute URL, this way it is easier for external aggregators to use the logo.

File: schema.md [hunk 13]
```diff
@@ -0,0 +1,555 @@
...
+
+This key contains the logo of the software. Logos should be in vector
+format, but raster formats are allowed as a fallback.
+
+Vector logos should be provided at least in two formats: a color one,
```
Comment (bvhme) [inline schema.md:305]: I only see the possibility for adding one logo, or should this be a list of strings?

File: schema.md [hunk 14]
```diff
@@ -0,0 +1,555 @@
...
+and a monochromatic (black) one.
+
+Raster logos should be provided only if vector logos do not exist.
+In this case, they should be transparent PNGs, minimum 1000px of
+width.
+
+### Key `description/shortdesc`
```
Comment (bvhme) [inline schema.md]: Inconsistent key naming (eg `repo-owner` and `shortdesc`)

Consider renaming this to kebab style `short-description` or renaming other things to `repoowner`.

File: schema.md [hunk 15]
```diff
@@ -0,0 +1,555 @@
...
+should be that of users of the software, not developers. You can think
+of this text as the description of the software that would be in its
+website (if the software had one).
+
+### Key `description/screenshots`
+
+* Type: array of strings (paths)
+* Presence: optional
+* Formats: PNG, JPG
+* Example: "data/screenshots/configuration.png"
```
Comment (bvhme) [inline schema.md:383]: Again, I'd make these absolute. Use GitHub CDN if necessary

File: schema.md [hunk 16]
```diff
@@ -0,0 +1,555 @@
...
+
+
+
+
+
+
+
+
+
+## Section `meta`
```
Comment (bvhme) [inline schema.md]: I'm wondering if `meta` might be the wrong word here. Since the description etc is also meta-data, and maybe even everything in this file.

May I suggest the word `context`?

File: schema.md [hunk 17]
```diff
@@ -0,0 +1,555 @@
...
+
+This section contains information useful to classify the software,
+define its scope of usage and its goal at a general level.
+
+### Key `meta/scope`
```
Comment (bvhme) [inline schema.md]: This should then perhaps be 'countries'

File: schema.md [hunk 18]
```diff
@@ -0,0 +1,555 @@
...
+
+* Type: string or array of strings
+* Presence: optional
+* Example: "es"
+
+This key defines the expected geographical scope of the software. If
+a software is probably useful only within some specific countries
+(because, eg., it helps implementing a specific regulation), specify
+the ISO693-1 two-letter country codes of the countries n this key.
```
Comment (bvhme) [inline schema.md:424]: ISO693-1 is not a standard for countries but for **languages**!

https://www.iso.org/standard/22109.html

File: schema.md [hunk 19]
```diff
@@ -0,0 +1,555 @@
...
+
+This key defines the types of public administration which is expected
+to use this software. Public software is often specific in scope, because
+there is a large set of tasks that are specific to each type of
+administration. For instance, many softwares that are used in schools
+are probably not useful in hospitals.
+
+The list of allowed values is defined in [pa-types.md](pa-types.md),
+and can evolve at any time, separately from the version of this
+specification.
```
Comment (bvhme) [inline schema.md:441]: This is a very complex one, and I'm looking forward to working on figuring this out a bit more, since even words like `school` or `city` vary wildly in meaning between contexts. But this is good for now!

File: pa-types.md [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+
+Value | Description
+----- | -----------
+`city` | City
+`hospital` | Hospital
+`police` | Police forces
+`school` | School of any level
+`university` | University of any level
+
+### Italy
```
Comment (bvhme) [inline pa-types.md:18]: Awesome, perhaps we can think of some method of extending the base set non-centrally.

File: schema.md [hunk 21]
```diff
@@ -0,0 +1,555 @@
...
+
+### Key `meta/category`
+
+* Type: enumerated string
+* Presence: optional
+* Values: see [sw-category.md](sw-category.md)
```
Comment (bvhme) [inline schema.md]: This file does not exist.

File: schema.md [hunk 22]
```diff
@@ -0,0 +1,555 @@
...
+for the software. You can select the category that is more similar to
+the software; it doesn't have to be an exact match, but it can still
+be very useful for building catalogs of software and evaluate
+alternatives.
+
+The list of allowed values is defined in [sw-category.md](sw-category.md),
+and can evolve at any time, separately from the version of this
+specification.
+
+### Key `meta/tags`
```
Comment (bvhme) [inline schema.md]: A bit more explanation on the difference between tags and categories would be useful.

File: schema.md [hunk 23]
```diff
@@ -0,0 +1,555 @@
...
+
+This section provides an overview on the system-level dependencies
+required to install and use this software.
+
+**NOTE:** do not list dependencies at the source code level (eg: software
+libraries being used), and focus only on runtime and/or system-level
+dependencies that must be installed and maintained separately. For
+instance, a database is a good example of such dependencies.
+
+### Key `section/open`
```
Comment (bvhme) [inline schema.md]: I think you mean `dependencies/open`

File: schema.md [hunk 24]
```diff
@@ -0,0 +1,555 @@
...
+* Presence: optional
+* Example: "MariaDB 10.2"
+
+This key contains a list of runtime dependencies that are distributed
+under an open-source license.
+
+Each string is free form, max 50 characters; feel free to add version
+numbers and/or short comments.
+
+### Key `section/proprietary`
```
Comment (bvhme) [inline schema.md]: I think you mean `dependencies/proprietary`

File: schema.md [hunk 25]
```diff
@@ -0,0 +1,555 @@
...
+* Presence: optional
+* Example: "IBM SoftLayer 4.5"
+
+This key contains a list of runtime dependencies that are distributed
+under a proprietary license.
+
+Each string is free form, max 50 characters; feel free to add version
+numbers and/or short comments.
+
+### Key `section/hardware`
```
Comment (bvhme) [inline schema.md]: I think you mean `dependencies/hardware`

File: schema.md [hunk 26]
```diff
@@ -0,0 +1,555 @@
...
+## Special data formats
+
+### Dates
+
+All dates in `publiccode.yml` must follow the format "YYYY-MM-DD",
+which is one of the ISO8601 allowed encoding. This is the only
+allowed encoding though, so not the full ISO8601 is allowed for the
+date keys.
+
+## Multi-language strings
```
Comment (bvhme) [inline schema.md:544]: I would still like to think about how to parse this a bit more, but for now this seems like a good idea to have in there.

**Review by bvhme — CHANGES_REQUESTED** [review]
Thanks :bow: this is some great work and it makes me super excited.

I do think there are some issues to resolve, that shouldn't be too hard before we get to a `0.1.0`. But this should be fairly straightforward.

The only thing I'm still a bit worried about which is fundamental at this stage is #18.

👋 Thanks and have a nice day!

**Comment by bvhme** [comment]
Also this PR solves point 1 from #16 by @sbronzolo.

File: schema.md (hunk 1 above)
Comment (rasky) [reply schema.md:534]: done

File: schema.md (hunk 2 above)
Comment (rasky) [reply schema.md:39]: I personally don't think full semver spec is relevant. I think what we want is to have a 1.0, evolve it and then maybe break it at some point in the distant future. So I think two numbers are enough.

File: schema.md (hunk 3 above)
Comment (rasky) [reply schema.md]: done

File: schema.md (hunk 4 above)
Comment (rasky) [reply schema.md]: The idea here is that the copyright could be diffuse. If you exclude CLA-based projects, if I send a single PR to a large open source project, I'm one of the copyright owners, but maybe not the main one. Sometimes there is no single "main" owner, but I think that being able to list something still makes sense. Linux uses "Linus Torvarld and all Linux contributors" which is OK IMO as information to show.

Noted `homepage` for next PR.

File: schema.md (hunk 5 above)
Comment (rasky) [reply schema.md]: Good idea.

File: schema.md (hunk 6 above)
Comment (rasky) [reply schema.md]: I tried a different approach:

> In case of a maintenance, this key must contain the date
at which the maintenance is going to end. In case of community
maintenance, the value should not be more than 2 years in the
future, and thus will need to be regularly updated as the
community continues working on the project.

File: schema.md (hunk 7 above)
Comment (rasky) [reply schema.md:177]: OK.

File: example/publiccode.yml (hunk 9 above)
Comment (rasky) [reply example/publiccode.yml]: YAML supports both formats. I usually put quotes only when required like in this case for the special characters.

File: schema.md (hunk 10 above)
Comment (rasky) [reply schema.md]: Done

File: schema.md (hunk 2 above)
Comment (bvhme) [reply schema.md:39]: Cool, should we not be at 0.1 though until we actually release this?

File: schema.md (hunk 4 above)
Comment (bvhme) [reply schema.md]: All right, I wonder if maybe `author` would be more precise then and perhaps a bit more explanatory?

What is the goal for this entry?

File: example/publiccode.yml (hunk 9 above)
Comment (bvhme) [reply example/publiccode.yml]: Awesome :)

**Comment by bvhme** [comment]
@rasky have you pushed your changes? I can't see them :)

File: schema.md (hunk 12 above)
Comment (rasky) [reply schema.md:300]: The URL has the following problems:

 * It changes under a maintained fork (a technical fork -- mad for the purpose of sending a PR -- should not touch publiccode.yml).
 * It encodes details of the code-hosting platform
 * It doesn't allow offline validation of a repo that has been cloned
 * It doesn't allow validation of a checkout at a specific point of Git history, as the URL would probably always point to the latest version of the file (which might not even exist anymore in the current master branch).

File: schema.md (hunk 13 above)
Comment (rasky) [reply schema.md:305]: Here and for other keys, I specified type as "strings or array of strings".

File: schema.md (hunk 14 above)
Comment (rasky) [reply schema.md]: As agreed in the issue, I'll send a separate PR to migrate to camelCase.

File: schema.md (hunk 16 above)
Comment (rasky) [reply schema.md]: OK.

File: schema.md (hunk 17 above)
Comment (rasky) [reply schema.md]: OK.

File: schema.md (hunk 18 above)
Comment (rasky) [reply schema.md:424]: Ops you're right: ISO-3166 is for countries. Do you have any preference for 2 letters vs 3 letters? Maybe 3 letters is more clear?

File: schema.md (hunk 21 above)
Comment (rasky) [reply schema.md]: Ah yes sorry, I'm still working on it.

File: schema.md (hunk 11 above)
Comment (rasky) [reply schema.md]: Done

File: schema.md (hunk 19 above)
Comment (rasky) [reply schema.md:441]: Sure we can evolve from here.

File: schema.md (hunk 22 above)
Comment (rasky) [reply schema.md]: Correct. We're thinking of this. Our current reasoning is that "category" is one or more tags from a closed list, while "tags" is a list of open tags.

File: schema.md (hunk 23 above)
Comment (rasky) [reply schema.md]: Done

File: schema.md (hunk 24 above)
Comment (rasky) [reply schema.md]: Done

File: schema.md (hunk 25 above)
Comment (rasky) [reply schema.md]: Done

**Comment by rasky** [comment]
@bvhme pushed now

**Comment by rasky** [comment]
Also pushed some slight update and camelCase conversion.

**Comment by bvhme** [comment]
I've closed and reopened this PR because I've added a 'develop' branch as the default. I'm keen to get this merged soon though! 

@rasky what would you like from me to get this merged?
